Ability to exclude links based on criteria

baldurrensch opened the pull request:
This is based on discussions in https://github.com/TheFootballSocialClub/FSCHateoasBundle/issues/34 and expands on the PR https://github.com/TheFootballSocialClub/FSCHateoasBundle/pull/36 by @marapper.

I have squashed his 2 commits into one. I have changed the annotation to `excludeIf`, the yml configuration value is `exclude_if`. It now is property path and a value. Example: 

``` php
/**
 * @Rest\Relation(
 *      "parent", 
 *      href = @Rest\Route("api_post_get", parameters = { "id" = ".parent.id"}), 
 *      excludeIf = { ".parent" = null } 
 * )
 */
```

or

``` yml
relations:
    - rel: parent
      href:
          route: api_post_get
          parameters: { id: .parent.id }
      exclude_if:
          ".parent": ~
```

marapper committed:
skipIfNull parameter in @Relation

Commit:
Renamed skip_if_null to exlude_if

- Now checks for the value and skips
- Added unit tests

adrienbrault commented:
What do you think about `exclude_if` ? Since the serializer already have a metadata with the same meaning

adrienbrault commented on Factory/LinkFactory.php:
```diff
@@ -43,16 +49,42 @@ public function createLinksFromMetadata(ClassMetadataInterface $classMetadata, $
+         */
         foreach ($classMetadata->getRelations() as $relationMetadata) {
-            $links[] = $this->createLinkFromMetadata($relationMetadata, $object);
+            if($link = $this->createLinkFromMetadata($relationMetadata, $object)) {
```
missing space after if

adrienbrault replied:
And why not check `$this->shouldSkipLink($relationMetadata, $object)` here ? It would avoid changing `createLinkFromMetadata`

baldurrensch replied:
Fixed

adrienbrault commented on Factory/LinkFactory.php:
```diff
@@ -43,16 +49,42 @@ public function createLinksFromMetadata(ClassMetadataInterface $classMetadata, $
 
+    protected function shouldSkipLink(RelationMetadataInterface $relationMetadata, $object)
+    {
+        if(! $fields = $relationMetadata->getSkipIf()) {
```
space thing here too

baldurrensch replied:
Fixed

adrienbrault commented on Factory/LinkFactory.php:
```diff
@@ -43,16 +49,42 @@ public function createLinksFromMetadata(ClassMetadataInterface $classMetadata, $
-        if (null !== $relationMetadata->getUrl()) {
+        if ($this->shouldSkipLink($relationMetadata, $object)) {
+            return null;
+        } elseif (null !== $relationMetadata->getUrl()) {
```
I think @stof would say that the else if can be just an if since the first if returns ... but I don't remember why :smiley_cat:

baldurrensch replied:
Fixed. @stof happy.

adrienbrault commented on Tests/Functional/TestBundle/Model/SkippedPost.php:
```diff
@@ -0,0 +1,26 @@
+{
+	protected $parent;
+
+	public static function create($id, $title, $parent = null)
```
weird indentation

baldurrensch replied:
Fixed

baldurrensch commented:
I think you are right, `exclude_if` is better, and more in line of what the serializer does. Changed. I also made all the other suggested changes, and rebased.

baldurrensch commented:
I will add documentation to the README before this gets merged in

Commit:
Updated README
